src/lazyapplying.jl

```diff
@@ -214,10 +217,14 @@ result_mul_style(::LazyArrayApplyStyle, _) = LazyArrayApplyStyle()
 result_mul_style(_, ::LazyArrayApplyStyle) = LazyArrayApplyStyle()
 
 
-struct  ApplyLayout{F, LAY} <: MemoryLayout end
+struct  ApplyLayout{F} <: MemoryLayout end
+
+applylayout(::F, args...) where F = ApplyLayout{F}()
 
-MemoryLayout(M::Type{Applied{Style,F,Args}}) where {Style,F,Args} = ApplyLayout{F,tuple_type_memorylayouts(Args)}()
-MemoryLayout(M::Type{ApplyArray{T,N,F,Args}}) where {T,N,F,Args} = ApplyLayout{F,tuple_type_memorylayouts(Args)}()
+MemoryLayout(::Type{Applied{Style,F,Args}}) where {Style,F,Args} = 
+    applylayout(F.instance,tuple_type_memorylayouts(Args).parameters...)
+MemoryLayout(::Type{ApplyArray{T,N,F,Args}}) where {T,N,F,Args} = 
+    applylayout(F.instance,tuple_type_memorylayouts(Args).parameters...)
```

Comment:
I guess `F.instance` would break for general callables? Though I can see that typing `::Type{typeof(+)}` etc. would be annoying. But maybe using `F` instead of `F.instance` with a helper shortcut function `TypeOf(f) = Type{typeof(f)}` is OK?

Reply:
Yes I agree this is less than ideal, and you are right that the motivation was just to make it less "annoying". It's probably also what's breaking type inference on 1.0.

I think the helper function is overkill since these are one-off implementations. 

I'll sleep on it... my only hesitation on reverting to the less clever version is it feels "harder" on the compiler, in some sense that's too early optimisation but the compiler seems to be quite delicate in these situations. (Though it sounds like its going to improve shortly.)

Reply:
Thinking it more, there’s plenty of types (like ` Fun ` ) that represent functions that cannot be instantiated. So I’ll change this to be less clever

Reply:
Thanks for making me look at this, as I also realised in the process the setup with `tuple_type_memorylayouts` was insane: it was creating a `Type{<:Tuple}` from a tuple, and then re-instantiating to get back the original tuple! It's now fixed to just pass the tuple (and hence the code isinferrable in Julia v1.0)